contracts/button/lib.rs

```diff
@@ -112,16 +162,16 @@ pub mod button_game {
             let access_control = AccountId::from(ACCESS_CONTROL_PUBKEY);
             let access_control = AccessControlRef::from_account_id(access_control);
 
-            match ButtonGame::check_role(&access_control, caller, required_role) {
-                Ok(_) => Self::init(
+            match access_control.has_role(caller, required_role) {
+                true => Self::init(
                     access_control,
                     ticket_token,
                     reward_token,
                     marketplace,
                     button_lifetime,
                     scoring,
                 ),
-                Err(why) => panic!("Could not initialize the contract {:?}", why),
+                false => panic!("Caller is not allowed to initialize this contract"),
             }
```

Comment:
nit: this can be replaced with more readable version IMO:
```rust
if !access_control.has_role(caller, required_role) {
  panic!("Caller is not allowed to initialize this contract")
}
Self::init(
                    access_control,
                    ticket_token,
                    reward_token,
                    marketplace,
                    button_lifetime,
                    scoring,
                )
```

and as a follow-up we might ticket adding a macro that checks the roles and panics if `false`.

Reply:
I'm a big fan of pattern matching - albeit which is limited in Rust when compared to [Clojure](https://github.com/clojure/core.match#example-usage), ymmv, I find it eons more readable than any branching logic.